diff --git a/CMakeLists.txt b/CMakeLists.txt
index f62f9df5..7bbf9fe3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -439,9 +439,9 @@ else()
     list(APPEND CMAKE_PREFIX_PATH ${OR_TOOLS_PATH})
   endif()
 
-  list(APPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_SOURCE_DIR}/ext)
+  # list(APPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_SOURCE_DIR}/ext)
 
-  find_package(ortools CONFIG)
+  find_package(ortools CONFIG REQUIRED)
 
   if (ortools_FOUND)
     get_filename_component(ortools_CMAKEDIR ${ortools_DIR} DIRECTORY)
